stellargraph/core/element_data.py

```diff
@@ -303,74 +303,117 @@ def _numpyise(d, dtype):
     return {k: np.array(v, dtype=dtype) for k, v in d.items()}
 
 
+class AdjList:
+    def __init__(self, init_index):
+        self.init_index = init_index
```

Comment:
Could you talk a bit about the design choice of passing in a function here? I'm not saying it's wrong and I'm not sure there's a _best_ answer at all, but I wonder if there's other versions that are a little clearer?

In particular, it feels a bit funny to have `EdgeData` store an `AdjList` object that holds a method from that parent `EdgeData`. An alternative might be something like

```python
class AdjList:
    def __init__(self, by_other_node_type, ...):
        if by_other_node_type:
            self._init_func = self._init_by_other_node_type
        else:
            self._init_func = self._init_plain

        self.undirected = self.ins = self.outs = None

    def _init_plain(self, ...):
        ...

    def _init_by_other_node_type(self, ...):
        ...
```

I guess we'll need to pass more arguments around in this case.

Reply:
Hmm yeah I wanted `AdjList` to handle the lookup + lazy initialising in one place, but thought it would be more convenient to define more adjacency-initialiser functions in `EdgeData` rather than in `AdjList` since I'm guessing it'll need access to all the type information in `EdgeData` and `NodeData` to build the adjacency lists with, so that's how I ended up with this approach. 

Not sure if you had something else in mind with your suggestion, but I guess we could pass in the whole edge data and node data into the adjacency list to do whatever it needs to do... but maybe that also looks a bit funny?
```python
class AdjList:
    def __init__(self, by_other_node_type, edge_data, node_data)
    def _init_plain(self, ...):
    def _init_by_other_node_type(self, ...):

class EdgeData:
    def __init__(self, ...):
        ...
        self._adj_list = AdjList(by_other_node_type=False, self, node_data)
        self._adj_list_by_other_node_type(by_other_node_type=True, self, node_data)
```

Reply:
Yeah, that's a bit funny too.

I thought about this overnight and I think my "concern" here is that the recursiveness of the data structure initialisation introduces risks of attempting to use a value while it's being initialised (e.g. if edge list initialisation happened to look at other edge lists, for some reason).

I guess the fully-separated version would be passing in `sources`, `targets` and `node_data`?

However, this may not be worth it, especially as edge list initialisation is:

- limited enough that there doesn't seem to be a reason why it would look at itself
- done as a final commit, rather than incremental construction (specifically, the `self.undirected` properties go from `None` to "fully ready", rather than being incrementally built up), and so we'll see infinite recursion here rather than invalid results

The main risk might be around delaying garbage collection (since this causes a cycle between objects, and so cPython will have to wait for the actual garbage collector to run, rather than objects being deallocated as soon as they go out of scope), and future refactoring that change either of the points above.